kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/MongoDebeziumTransform.java

```diff
@@ -0,0 +1,334 @@
+        record.kafkaPartition(),
+        newKeySchema,
+        newKeyStruct,
+        newValueSchema,
+        newValueStruct,
+        record.timestamp(),
+        record.headers());
+  }
+
+  private SchemaBuilder mutateBuilderFromBson(
+      SchemaBuilder builder, BsonDocument bson, String fieldName) {
+    SchemaBuilder innerBuilder = SchemaBuilder.struct();
+    Set<Map.Entry<String, BsonValue>> pairs = bson.entrySet();
+    for (Map.Entry<String, BsonValue> pairsForSchema : pairs) {
+      converter.addFieldSchema(pairsForSchema, innerBuilder);
+    }
+    builder.field(fieldName, innerBuilder.optional().build());
+    return builder;
+  }
+
+  private Struct fillStructFromBson(Schema schema, BsonDocument bson) {
+    Struct struct = new Struct(schema);
+    Set<Map.Entry<String, BsonValue>> pairs = bson.entrySet();
+    for (Map.Entry<String, BsonValue> pairsForSchema : pairs) {
+      converter.convertRecord(pairsForSchema, schema, struct);
+    }
+    return struct;
+  }
+
+  private boolean isValidKey(final R record) {
+    if (record.keySchema() == null
+        || record.keySchema().name() == null
+        || !record.keySchema().name().endsWith(RECORD_ENVELOPE_KEY_SCHEMA_NAME_SUFFIX)) {
+      LOG.debug(
+          "Expected Key Schema for transformation, converting to tombstone. Message key: \"{}\"",
+          record.key());
+      return false;
+    }
+    return true;
+  }
+
+  private boolean isValidValue(final R record) {
+    if (record.valueSchema() == null
+        || record.valueSchema().name() == null
+        || !containsEnvelopePrefix(record.valueSchema())) {
+      LOG.debug("Expected Envelope for transformation, converting to tombstone");
+      return false;
+    }
+    return true;
+  }
+
+  private boolean containsEnvelopePrefix(Schema schema) {
+    return schema.name().endsWith(SCHEMA_NAME_SUFFIX);
+  }
+
+  private static <R extends ConnectRecord<R>> ExtractField<R> extractorValueField(String field) {
+    ExtractField<R> extractField = new ExtractField.Value<>();
+    Map<String, String> target = Maps.newHashMap();
+    target.put("field", field);
+    extractField.configure(target);
+    return extractField;
+  }
+
+  private static <R extends ConnectRecord<R>> ExtractField<R> extractorForKeyField(String field) {
+    ExtractField<R> extractField = new ExtractField.Key<>();
+    Map<String, String> target = Maps.newHashMap();
+    target.put("field", field);
+    extractField.configure(target);
+    return extractField;
+  }
+
+  private String kafkaMetadataForException(R record) {
+    Long kafkaOffset = null;
+    try {
+      SinkRecord sinkrecord = (SinkRecord) record;
+      kafkaOffset = sinkrecord.kafkaOffset();
+    } catch (Exception ignored) {
+      return String.format("topic %s, partition: %s", record.topic(), record.kafkaPartition());
+    }
+    return String.format(
+        "topic: %s, partition: %s, offset: %s",
+        record.topic(), record.kafkaPartition(), kafkaOffset);
+  }
```

Comment:
Once you have just a SinkRecord (instead of R), then this can be simplified to: 

```suggestion
  private String kafkaMetadataForException(SinkRecord sinkrecord) {
    return String.format(
            "topic: %s, partition: %s, offset: %s",
            sinkrecord.topic(), sinkrecord.kafkaPartition(), sinkrecord.kafkaOffset());
  }
```